### PR TITLE
feat: Add nomad fmt

### PR DESCRIPTION
**Pull request by Trojan295**
This PR adds the `nomad fmt` command, which can be used to format and check Nomad configuration and jobspec files.

It formats the files according to the HCL syntax using [hashicorp/hcl](https://github.com/hashicorp/hcl/). It does not validate, if the files are correct Nomad or Jobspec files.

Closes https://github.com/hashicorp/nomad/issues/11757

### CODE REVIEW

File: command/fmt.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	with .nomad and .hcl extensions in the directory.
+
+Format Options:
+
+	-check
+		Check if the files are valid HCL files. If not, exit status of the command
+		will be 1 and the incorrect files will not be formatted.
+
+	-overwrite=false
+		Print the formatted files to stdout, instead of overwriting the file content.
```
Comment (tgross) [inline command/fmt.go:82]: Let's call this `write` to match `terraform fmt` (ref https://developer.hashicorp.com/terraform/cli/commands/fmt#write-false).

File: command/fmt.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	flags := f.Meta.FlagSet(f.Name(), FlagSetClient)
+	flags.Usage = func() { f.Ui.Output(f.Help()) }
+	flags.BoolVar(&fmtArgs.Check, "check", false, "")
+	flags.BoolVar(&fmtArgs.Overwrite, "overwrite", true, "")
+
+	if err := flags.Parse(args); err != nil {
+		return 1
+	}
+
+	fmtArgs.Paths = flags.Args()
```
Comment (tgross) [inline command/fmt.go:119]: As my colleague @mikenomitch noted in https://github.com/hashicorp/nomad/issues/11757#issuecomment-1266090490, let's also allow this to accept `-` to read files from stdin.

File: command/fmt.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	return 0
+}
+
+func (*FormatCommand) Help() string {
+	helpText := `
+Usage: nomad fmt [flags] paths ...
+
+	Formats Nomad agent configuration and job file to a canonical format.
+	If a path is a directory, it will recursively format all files
+	with .nomad and .hcl extensions in the directory.
```
Comment (tgross) [inline command/fmt.go:73]: `terraform fmt` doesn't do this by default and has a separate [-recursive](https://developer.hashicorp.com/terraform/cli/commands/fmt#recursive) option. We should probably do the same.

File: command/fmt_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	ui := cli.NewMockUi()
+
+	cmd := &FormatCommand{
+		Meta: Meta{Ui: ui},
+	}
+
+	bytes, err := os.ReadFile("testdata/fmt/bad-nomad.hcl")
+	require.NoError(t, err)
+
+	f, err := os.CreateTemp("", "nomad-fmt-")
```
Comment (tgross) [inline command/fmt_test.go:66]: This leaves temp files on the test host when we're done. You can use [`t.TempDir`](https://pkg.go.dev/testing#T.TempDir) to get a directory that'll automatically get cleaned up afterwards.

**Review by tgross — CHANGES_REQUESTED** [review]
Hi @Trojan295! This is great!

I've left some comments that can tighten this up and make it more consistent with both `terraform fmt` and the rest of the Nomad code base. Two more things:

* You can add a changelog file. You'll make that at `.changelog/14779.txt` and it'll look like https://github.com/hashicorp/nomad/blob/main/.changelog/14742.txt. 
* There's documentation at [./website/content/docs/commands](https://github.com/hashicorp/nomad/tree/main/website/content/docs/commands). You can take a crack at mimicking the files in the top of that directory, but if that's too overwhelming just let me know and I can knock that out pretty quickly once this is ready to merge.

Thanks again!

File: command/fmt.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+func (*FormatCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictFiles("*")
+}
+
+func (*FormatCommand) AutocompleteFlags() complete.Flags {
+	return complete.Flags{
+		"-check":           complete.PredictNothing,
+		"-overwrite=false": complete.PredictNothing,
```
[the remark this answers is not in the document]
Comment (Trojan295) [reply command/fmt.go:99]: My idea behind this was, that by default the `overwrite` defaults to `true`, so someone would provide the flag to change this flag to `false`. That's why I think it's easier, if this autocompletes to `false` already.

[the remark this answers is not in the document]
Comment (tgross) [reply command/fmt.go:99]: Ah yeah, I misread the docstring here.

File: website/content/docs/commands/fmt.mdx [hunk 6]
```diff
@@ -0,0 +1,58 @@
+---
```
Comment (tgross) [inline website/content/docs/commands/fmt.mdx:1]: I forgot that we need to add this page to the navbar for it to be rendered. If you add this bit of JSON:

```
      {
        "title": "fmt",
        "path": "commands/fmt"
      },
```

between the "commands/eval" and "commands/jobs" at https://github.com/hashicorp/nomad/blob/main/website/data/docs-nav-data.json#L431-L432, that'll allow our preview build for the docs to work.

**Review by tgross — CHANGES_REQUESTED** [review]
Looking great @Trojan295! Just a few last items to fix up and we'll be able to land this one.

**Comment by Trojan295** [comment]
Thanks @tgross for the review! I will update this PR tomorrow. :)

File: command/testdata/fmt/nomad.in.hcl [hunk 7]
```diff
@@ -0,0 +1,8 @@
+server {
```
[the remark this answers is not in the document]
Comment (Trojan295) [reply command/testdata/fmt/nomad.in.hcl:1]: What do you think about ignoring the `./command/testdata` directory from the `make hclfmt`?
```
   -o -path './command/testdata' -prune \
```

[the remark this answers is not in the document]
Comment (tgross) [reply command/testdata/fmt/nomad.in.hcl:1]: Eh, that'll work.

**Review by tgross — APPROVED** [review]
Looks great, @Trojan295! Thanks for sticking with it, I know PRs for new commands can feel like a whole lot of nitpicking 😀 

CircleCI seems to have gotten confused about the PR and can't run the build. So I just rebased this branch on `main` to get that moving again. Once that finishes, we can get this merged. ~(I'll probably have to do that tomorrow morning, as I'm close to end-of-day here.)~ _Done_

**Comment by tgross** [comment]
@Trojan295 this has been merged and a backport bot will merge it to the 1.4.x release branch for the next patch release of Nomad . If that bot can't figure things out and requests a review from you, don't worry about it and I'll fix it in the morning; it's one of those fiddly process things we try not to expose to contributors but the bot is working thru a bit of a queue at the moment and I won't be there to immediately catch it. 😀

**Comment by Trojan295** [comment]
> Looks great, @Trojan295! Thanks for sticking with it, I know PRs for new commands can feel like a whole lot of nitpicking 😀

Wasn't so bad :). I just had a though, that it would be nice, if some of those docs/CLI help could be generated. E.g. we have the help in `mdx` and this gets automatically added to the website and also the CLI help is generated out of it.

**Comment by tgross** [comment]
> I just had a though, that it would be nice, if some of those docs/CLI help could be generated.

Yes! I know a few folks have been tinkering around with better CLI generation internally for that reason.

**Pull request merged**